node/src/components/block_validator.rs

```diff
@@ -202,15 +219,57 @@ where
                 });
             }
             Event::DeployMissing(deploy_hash) => {
+                info!(%deploy_hash, "request to download deploy timed out");
                 // A deploy failed to fetch. If there is still hope (i.e. other outstanding
                 // requests), we just ignore this little accident.
                 if self.in_flight.dec(&deploy_hash) != 0 {
                     return Effects::new();
                 }
 
-                // Otherwise notify everyone still waiting on it that all is lost.
+                // Flag indicating whether we've retried fetching the deploy.
+                let mut retried = false;
+
                 self.validation_states.retain(|key, state| {
                     if state.missing_deploys.contains(&deploy_hash) {
+                        match state.source() {
+                            Some(peer) => {
+                                info!(%deploy_hash, ?peer, "trying the next peer");
+                                // There's still hope to download the deploy.
+                                let (chainspec, block_timestamp) = &state.context;
+                                effects.extend(fetch_deploy(effect_builder, Arc::clone(chainspec), *block_timestamp, deploy_hash, peer));
+                                retried = true;
+                                true
+                            },
+                            None => {
+                                // Notify everyone still waiting on it that all is lost.
+                                info!(block=?key, %deploy_hash, "could not validate the deploy. block is invalid");
+                                // This validation state contains a failed deploy hash, it can never
+                                // succeed.
+                                state.responders.drain(..).for_each(|responder| {
+                                    effects.extend(responder.respond((false, key.clone())).ignore());
+                                });
+                                false
+                            }
+                        }
+                    } else {
+                        true
```

Comment:
This would be much easier to read as an early return.

Reply:
How would that look like? We're iterating over all collection of `self.validation_states`.

Reply:
But we're inside of a closure here, so you can just:
```rust
if !state.missing_deploys.contains(&deploy_hash) {
    return true;
}
```

Reply:
I will revert the boolean and change the order of `if { … } else { … }` but I am not a fan of early return.

Reply:
I had to add yet another `if` so I added an early return this time.